pyquil/paulis.py

```diff
@@ -931,6 +932,63 @@ def combined_exp_wrap(param: Union[float, MemoryReference]) -> Program:
     return combined_exp_wrap
 
 
+def exponentiate_pauli_sum(
+    pauli_sum: PauliSum,
+) -> np.ndarray:
+    r"""
+    Exponentiates a sequence of PauliTerms, which may or may not commute. The Pauliterms must
+    have fixed (non-parametric) coefficients. The coefficients are interpreted in cycles
+    rather than radians or degrees.
+
+    e^{-i\pi\sum_i \theta_i P_i}
+
+    Thus, 1.0 corresponds to one full rotation.
+
+    To produce a CZ gate:
+
+    >>> phi = pi
+    >>> coeff = phi/(-4*pi) # -0.25
+    >>> hamiltonian = PauliTerm("Z", 0) * PauliTerm("Z", 1) - 1*PauliTerm("Z", 0) - 1*PauliTerm("Z", 1)
+    >>> exponentiate_pauli_sum(coeff*hamiltonian)
+
+    To produce the Quil XY(theta) gate, you can use:
+
+    >>> coeff = theta/(-4*pi)
+    >>> hamiltonian = PauliTerm("X", 0) * PauliTerm("X", 1) + PauliTerm("Y", 0) * PauliTerm("Y", 1)
+    >>> exponentiate_pauli_sum(coeff*hamiltonian)
+
+    A global phase is applied to the unitary such that the [0,0] element is always real.
+
+    :param pauli_sum: PauliSum to exponentiate.
+    :returns: The matrix exponetial of the PauliSum
+    """
+    if isinstance(pauli_sum, PauliTerm):
+        pauli_sum = PauliSum([pauli_sum])
+
+    pauli_matrices = {
+        "X": np.array([[0.0, 1.0], [1.0, 0.0]]),
+        "Y": np.array([[0.0, 0.0 - 1.0j], [0.0 + 1.0j, 0.0]]),
+        "Z": np.array([[1.0, 0.0], [0.0, -1.0]]),
+        "I": np.array([[1.0, 0.0], [0.0, 1.0]]),
+    }
+
+    qubits = pauli_sum.get_qubits()
+    for q in qubits:
+        assert isinstance(q, int)
+    qubits.sort()
+
+    matrices = []
+    for term in pauli_sum.terms:
+        coeff = term.coefficient
+        assert isinstance(coeff, (int, float, complex))
+        qubit_paulis = {qubit: pauli for qubit, pauli in term.operations_as_set()}
+        paulis = [qubit_paulis[q] if q in qubit_paulis else "I" for q in qubits]
+        matrix = float(np.real(coeff)) * reduce(np.kron, [pauli_matrices[p] for p in paulis])
+        matrices.append(matrix)
+    generated_unitary = expm(-1j * np.pi * sum(matrices))
+    return np.exp(-1j * np.angle(generated_unitary[0, 0])) * generated_unitary
```

Comment:
Apropos of

> I don't know how to fix this error:

maybe pull it out and `.astype` it, or throw in the ol' trusty `# type: ignore`?

```suggestion
    angle = np.angle(generated_unitary[0, 0])  # type: ignore
    return np.exp(-1j * angle) * generated_unitary
```

Reply:
You could also beef up the return type here; I realize this is old code and elsewhere there's `np.ndarray` as a return type, so it's not just you, but maybe something like

```python
from numpy.typing import NDArray

…

def exponentiate_pauli_sum(pauli_sum: PauliSum) -> NDarray[np.complex128]:
…
```